### PR TITLE
[CSHARP-969] - OpenTelemetry session level tracing

### PR DESCRIPTION
**Pull request by simaoribeiro**
As discussed in the [OpenTelemetry proposal](https://github.com/datastax/csharp-driver/blob/master/proposals/open-telemetry/tracing-rfc.md), this PR includes the session level tracing following the OpenTelemetry standard.

@joao-r-reis There are two things that are included in the proposal that are missing in this:
- The `server.address` and `server.port` attributes because the Host information is only set in the Node level tracker methods
- The operation name, as this is a more of a big deal, as the `RequestTrackingInfo` is instantiated in the `RequestHandler` but the method invocation is made before in the `Session` class. One solution is to instantiate the tracker at the session level so we can access the method, but I wanted to see your opinion before going for that route in this PR.

### CODE REVIEW

**Comment by simaoribeiro** [comment]
Hi @joao-r-reis , a bit later than I was expecting but here is the PR so we can proceed with the feature :)

**Comment by joao-r-reis** [comment]
> The operation name, as this is a more of a big deal, as the RequestTrackingInfo is instantiated in the RequestHandler but the method invocation is made before in the Session class. One solution is to instantiate the tracker at the session level so we can access the method, but I wanted to see your opinion before going for that route in this PR.

We didn't really cover how we were going to implement this part and now that I'm thinking a bit more about it we would have to add a new `Enum` or just pass a string in the `TrackingInfo` but it's a bit awkward because it really only would be useful for this specific implementation of the `RequestTracker`. 

At the same time I'm thinking whether this will actually be useful and what would we lose if we just made the OpenTelemetry extension hardcode the operation name to `Execute`... If we were to implement the new "field" on the tracker info object at best we would have 2 operation names: `Execute` and `ExecuteAsync` but then there's some cases where LINQ and Mapper call `ExecuteAsync` even if the top level method is a synchronous one so I'm actually leaning towards just hardcoding a string like `CassandraRequest` as the operation name in the extension (it should be a non existing driver method to avoid confusion I think)... And maybe in the future we can revisit this and potentially add a new field to the tracking info object (which wouldn't be a breaking change).

---

I can't do a proper review of the PR right now but I'll try finding some time soon to look at it, thanks for the contribution again!

**Comment by zehenrique** [comment]
hey @joao-r-reis how's it going?

do you think you will be able to look at this soon? This particular subject is becoming an issue and we need to understand how we will move forward in the near future.

thanks

**Comment by joao-r-reis** [comment]
> hey @joao-r-reis how's it going?
> 
> do you think you will be able to look at this soon? This particular subject is becoming an issue and we need to understand how we will move forward in the near future.
> 
> thanks

Hey, I'll do a proper review today but it was my impression that this was still a draft and there was still work left to be done.

**Comment by simaoribeiro** [comment]
Hi @joao-r-reis , it was my bad. I didn't realized it was marked as draft. I believe this pull request implements what we discussed in technical document minus the operation name.

Regarding that, I agree that we can change it to something more generic like the hardcoded string you mentioned above.

File: src/Cassandra/Requests/RequestExecution.cs
Comment (joao-r-reis) [inline src/Cassandra/Requests/RequestExecution.cs:392]: By the way, you can add the `OnNodeSuccess` call at the top of the `private void HandleRowSetResult(Response response)` method and with this we will have all the request tracker methods implemented in the core driver 👍

**Review by joao-r-reis — COMMENTED** [review]
Looking really good! Added a few comments.

I think initially we discussed that the `OnNodeSuccess` and `OnNodeFailure` methods were more complex to implement so they were going to be out of scope for now but it turns out they are super easy to implement after all. The `OnNodeFailure` is pretty much already implemented in your changes and I added a comment that says where you can add the `OnNodeSuccess` call.

**Comment by joao-r-reis** [comment]
The build is still failing due to some analyzers that are not enabled by default when you build locally. If you want to build with all the analyzers enabled locally you can follow [these instructions](https://github.com/datastax/csharp-driver/blob/master/CONTRIBUTING.md#automated-checks---code-analyzers).

File: src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs
Comment (joao-r-reis) [inline src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs:14]: Can we have a test that verifies that the activity context is correctly propagated? I don't think there is one at the moment that verifies this unless I missed it. If I'm not mistaken one of the use cases for this is to properly correlate these traces with (for example) an API call so the context needs to be properly propagated correct? Ideally we would have a test for statements, one for the mapper and one for LINQ. And each test should use both sync and async variants of the execute methods.

We should also test that the context propagation works correctly with auto paging, although I'm not 100% clear what behavior we want in this case. I'm assuming the context we want is the one that is active the moment where the enumerator pages the next page so the current changes might already do that but we need a test for it.

**Comment by simaoribeiro** [comment]
Hi @joao-r-reis , back at it again after some time off. Will start working on your review

File: src/Extensions/Cassandra.OpenTelemetry/CassandraInstrumentation.cs
Comment (simaoribeiro) [reply src/Extensions/Cassandra.OpenTelemetry/CassandraInstrumentation.cs]: Are we expecting it to change without being included in a major version? 🤔 

I followed the same logic as the OpenTelemetry contribution packages, e.g.: [SqlClient](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlActivitySourceHelper.cs#L19)

We also followed the same logic, last year, for our open-source [KafkaFlow](https://github.com/Farfetch/kafkaflow/blob/master/src/KafkaFlow.OpenTelemetry/ActivitySourceAccessor.cs) library and we didn't found any problems with the approach.

However, if you think its safer, I can change it to a hardcoded string without problems

File: src/Cassandra/OpenTelemetry/TracesRequestObserver.cs
Comment (simaoribeiro) [reply src/Cassandra/OpenTelemetry/TracesRequestObserver.cs:23]: As I'm on it, I will also include a folder for the Null and Composite observers, if it's ok for you

File: src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs
Comment (simaoribeiro) [reply src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs:14]: I agree. The context propagation tests can be straightforward. I will include a method that will be called after the database requests and check if the traceId is the same and also that the database span id is the parent of the method span id.

File: src/Cassandra/OpenTelemetry/TracesRequestObserver.cs
Comment (joao-r-reis) [reply src/Cassandra/OpenTelemetry/TracesRequestObserver.cs:23]: yeah sure

File: src/Extensions/Cassandra.OpenTelemetry/CassandraInstrumentation.cs
Comment (joao-r-reis) [reply src/Extensions/Cassandra.OpenTelemetry/CassandraInstrumentation.cs]: Ok let's keep it as is then

File: src/Extensions/Cassandra.OpenTelemetry/Implementation/Trace.cs
Comment (simaoribeiro) [reply src/Extensions/Cassandra.OpenTelemetry/Implementation/Trace.cs]: Done, please review

File: src/Extensions/Cassandra.OpenTelemetry/Implementation/Trace.cs
Comment (simaoribeiro) [reply src/Extensions/Cassandra.OpenTelemetry/Implementation/Trace.cs]: Done, please review

File: src/Extensions/Cassandra.OpenTelemetry/BuilderExtensions.cs
Comment (simaoribeiro) [reply src/Extensions/Cassandra.OpenTelemetry/BuilderExtensions.cs:44]: Done, please review

File: src/Cassandra/Requests/RequestExecution.cs
Comment (simaoribeiro) [reply src/Cassandra/Requests/RequestExecution.cs]: @joao-r-reis , I implemented this as you suggested and I also included a method named `OnNodeStart()` in the `RequestExecution.Start()`. 

The idea is to have nested activities for node level requests (that will appear inside the session level requests) that will give the user information regarding the host and execution retries.
As we already had the session start, error, and finish implementations, I think it didn't make much sense to have only the error and finish implementations for node requests as, for activities, there was nothing being started. With that in mind, the activity names are now called `Session Request` and `Node Request` instead of just `Request` as you suggested here.
https://github.com/datastax/csharp-driver/pull/608#issuecomment-2122837103

What do you think about this?

File: src/Cassandra/OpenTelemetry/TracesRequestObserver.cs
Comment (simaoribeiro) [reply src/Cassandra/OpenTelemetry/TracesRequestObserver.cs]: Done, please review

File: src/Cassandra/OpenTelemetry/TracesRequestObserver.cs
Comment (simaoribeiro) [reply src/Cassandra/OpenTelemetry/TracesRequestObserver.cs]: Done, please review

File: src/Cassandra/Builder.cs
Comment (simaoribeiro) [reply src/Cassandra/Builder.cs]: Done, please review

File: src/Cassandra/IRequestTracker.cs
Comment (simaoribeiro) [reply src/Cassandra/IRequestTracker.cs:25]: Done, please review

File: src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs
Comment (simaoribeiro) [reply src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs:14]: Working on the tests currently 👍

File: src/Extensions/Cassandra.OpenTelemetry/Implementation/OpenTelemetryRequestTracker.cs
Comment (joao-r-reis) [inline src/Extensions/Cassandra.OpenTelemetry/Implementation/OpenTelemetryRequestTracker.cs:21]: I don't think the `Implementation` folder/namespace adds much value here, we can just put this class on the "root" directory, also makes it easier for users without an IDE that auto suggests namespace imports.

File: src/Cassandra/Requests/RequestExecution.cs
Comment (joao-r-reis) [reply src/Cassandra/Requests/RequestExecution.cs]: Makes perfect sense 👍 I like it.

File: src/Cassandra/Observers/RequestTracker/RequestTrackerObserver.cs
Comment (joao-r-reis) [inline src/Cassandra/Observers/RequestTracker/RequestTrackerObserver.cs:42]: There's a minor optimization that can be done here, I know this was true some years ago but I don't know if modern .NET versions have made this irrelevant:

In cases like this where there's only one `await` and it's the last "statement" of the method then we can just remove `async await` and just return the task itself.

```
public Task OnNodeRequestError(
            Host host,
            RequestErrorType errorType,
            RetryDecision.RetryDecisionType decision,
            RequestTrackingInfo r,
            Exception ex)
        {
            var hostInfo = new HostTrackingInfo { Host = host };

            return _requestTracker.OnNodeErrorAsync(r, hostInfo, ex);
```

Comment (joao-r-reis) [reply src/Cassandra/Observers/RequestTracker/RequestTrackerObserver.cs:42]: This can also be done for the other methods of this class. Again it's a very minor thing and not that important but it does make it so there's one less `async/await` state machine that the compiler generates (unless modern .NET versions make this optimization automatically).

File: src/Cassandra/Observers/Abstractions/IRequestObserver.cs
Comment (joao-r-reis) [inline src/Cassandra/Observers/Abstractions/IRequestObserver.cs:40]: Should probably add `Async` suffix to the names of these methods to make it consistent.

File: src/Cassandra/Builder.cs
Comment (joao-r-reis) [reply src/Cassandra/Builder.cs]: This is still not addressed I think.

File: src/Cassandra/RequestTrackingInfo.cs
Comment (joao-r-reis) [inline src/Cassandra/RequestTrackingInfo.cs:30]: Do we need a public `setter` here? Ideally the driver should set this when it creates this object and these properties should be immutable (well the `Items` dictionary can be modified but the properties themselves shouldn't).

File: src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs
Comment (joao-r-reis) [reply src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs:14]: Let me know when I can take a look at the tests 👍

**Review by joao-r-reis — COMMENTED** [review]
Looks really good, I think most discussions are resolved now, there's only a couple of things left to address. IIUC you're still working on the tests right? I can take another look once that's done.

File: src/Cassandra/Requests/RequestExecution.cs
Comment (joao-r-reis) [inline src/Cassandra/Requests/RequestExecution.cs:399]: Hmm we might need to ensure this gets called before `_parent.SetCompleted` because otherwise the "OnRequestFailure" will get called before "OnNodeRequestError" on the request tracker...

And if the retry decision is `Retry` then the new "child trace" will be started before this one ends.

```
switch (retryInformation.Decision.DecisionType)
    {
        case RetryDecision.RetryDecisionType.Rethrow:
           _requestObserver.OnNodeRequestError(host, retryInformation.Reason, RetryDecision.RetryDecisionType.Rethrow, _requestTrackingInfo, ex);
            _parent.SetCompleted(ex);
            break;
        case RetryDecision.RetryDecisionType.Ignore:
            _requestObserver.OnNodeRequestError(host, retryInformation.Reason, RetryDecision.RetryDecisionType.Ignore, _requestTrackingInfo, ex);
            // The error was ignored by the RetryPolicy, return an empty rowset
            _parent.SetCompleted(null, FillRowSet(RowSet.Empty(), null));
            break;
        case RetryDecision.RetryDecisionType.Retry:
            _requestObserver.OnNodeRequestError(host, retryInformation.Reason, RetryDecision.RetryDecisionType.Retry, _requestTrackingInfo, ex);
            //Retry the Request using the new consistency level
            Retry(retryInformation.Decision.RetryConsistencyLevel, retryInformation.Decision.UseCurrentHost, host);
            break;
        default: 
           _requestObserver.OnNodeRequestError(host, retryInformation.Reason, retryInformation.Decision.DecisionType, _requestTrackingInfo, ex);
    }
```

The above isn't pretty but I don't think there's pretty and efficient solutions for this

File: src/Cassandra/Builder.cs
Comment (simaoribeiro) [reply src/Cassandra/Builder.cs]: Done now.

File: src/Extensions/Cassandra.OpenTelemetry/Implementation/OpenTelemetryRequestTracker.cs
Comment (simaoribeiro) [reply src/Extensions/Cassandra.OpenTelemetry/Implementation/OpenTelemetryRequestTracker.cs]: Yeah, I was following the https://github.com/open-telemetry/opentelemetry-dotnet-contrib structure as a guide but it doesn't make much sense in our implementation. I changed it in the last commit

File: src/Cassandra/Observers/RequestTracker/RequestTrackerObserver.cs
Comment (simaoribeiro) [reply src/Cassandra/Observers/RequestTracker/RequestTrackerObserver.cs]: I honestly don't know if the latest .NET versions have made the optimization but given that the package is .NET Standard, which can be executed by a .NET Framework runtime, I think it's safer to make the optimization.
Changed it in the last commit

File: src/Cassandra/Observers/Abstractions/IRequestObserver.cs
Comment (simaoribeiro) [reply src/Cassandra/Observers/Abstractions/IRequestObserver.cs]: Done, please review

File: src/Cassandra/RequestTrackingInfo.cs
Comment (simaoribeiro) [reply src/Cassandra/RequestTrackingInfo.cs]: I changed it so the statement is passed by constructor and the public property is getter only

File: src/Cassandra/Requests/RequestExecution.cs
Comment (simaoribeiro) [reply src/Cassandra/Requests/RequestExecution.cs]: Changed it in the last commit. Please review

File: src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs
Comment (simaoribeiro) [reply src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs:35]: Currently I have the following tests:

- When the Keyspace is available check if DbOperation and DbName attributes are included;
- When the Keyspace is not available available check if DbName attribute is not included;
- With default builder options, the statement is not included as an attribute ;
- With the option `IncludeDatabaseStatement` the statement is included as an attribute;
- A test with `Execute` and `Execute` async methods in which I verify if the trace is created and also if the node span is child of session span;
- A test with `Mapper` and `Mapper` async methods in which I verify if the trace is created and also if the node span is child of session span;

Currently missing is the LINQ test and the context propagation that you mentioned above.
I also wanna make a test with a retry to see if multiple spans are generated that are child of the same session span.

I will try to finish in the next few days so, if you have time available, you can start checking the current tests :)

Comment (simaoribeiro) [reply src/Cassandra.IntegrationTests/OpenTelemetry/OpenTelemetryTests.cs:39]: @joao-r-reis I finished the tests. Since my previous comment I included
- A test to validate the activity when executing using LINQ
- A test to validate if the same TraceId is used when invoking a different method with a call to Cassandra in the mix
- A test to validate if, when a host is down and a retry happens, if it is produced more than one node activity for the same session activity. During this test, I realized that if multiple retries happens on the same host, the node level activity would be reused instead of being created a new one, so I fixed it in this commit https://github.com/datastax/csharp-driver/pull/608/commits/13f3eec6001b1a5cc74f5cf155445d93037d0b6a
- A test for pagination

**Comment by simaoribeiro** [comment]
There are 3 tests failing regarding to Metrics, I'm still trying to understand why. Anyways, I made a rebase against master to get the last changes and I the commit times ended up a bit messy, but the commits since your last review starts on this one:
https://github.com/datastax/csharp-driver/pull/608/commits/5555e417bc316ec538f2a94fc826ffdcaf81e8b7

**Comment by joao-r-reis** [comment]
Thanks! I'm just wrapping up #611 and I'll take another look at this afterwards.

**Comment by joao-r-reis** [comment]
I'll start taking a look at this now, I'll try to figure out why the tests are failing too

**Comment by joao-r-reis** [comment]
So the issue is that the request observer methods are not being awaited. [I made the changes to support this and the changes turned out to be quite significant](https://github.com/datastax/csharp-driver/commit/b28863ed2898dbe2e5276e45c0f9707bd6eb9345). I'm running a quick benchmark to see if performance is impacted too much by this, if it does then we can just make the request observer (and request tracker) methods synchronous since they aren't needed by the open telemetry extension.

**Comment by joao-r-reis** [comment]
Oh there's a separate issue, there is no `CompositeConnectionObserver`, so the connection observer that handles node metrics is not being used. I'm also fixing that on a separate branch and I'm benchmarking these changes, I'll open a PR that targets this branch afterwards.

**Comment by joao-r-reis** [comment]
Can you add LINQ and Mapper test cases to the following test? `AddOpenTelemetry_WhenMethodIsInvokedAfterQuery_TraceIdIsTheSameThroughRequest`

Just to ensure the parent trace id is properly propagated even when using the mapper and linq.

Also add async to this test too to ensure that this propagation works with Statements,Mapper,LINQ AND both sync and async.

Everything else looks really good, I'm good to merge after this test is modified (or new test(s) is(are) added for these test cases that I mentioned).

**Comment by joao-r-reis** [comment]
[Opened a PR against your branch ](https://github.com/simaoribeiro/csharp-driver/pull/3)to fix the await calls.

**Comment by simaoribeiro** [comment]
Now makes sense why the tests were failing. I was having the problem of the metrics tests failing locally both on the master and opentelemetry branches so I wasn't sure why. Thanks for looking at it.

I'm adding the use cases you mentioned for `AddOpenTelemetry_WhenMethodIsInvokedAfterQuery_TraceIdIsTheSameThroughRequest`. I think I will have it finished by tomorrow afternoon so we can proceed with the merge.

Do you want me to write (or help you write) documentation regarding the feature, after this?

**Comment by joao-r-reis** [comment]
Sorry @simaoribeiro I was about to write that I added the test case I mentioned to the PR I just opened against your branch.

> Do you want me to write (or help you write) documentation regarding the feature, after this?

Yeah, after my PR is merged to your branch I think that all that is left is the documentation. I'm planning on spending the next couple of days setting up a local env to test this new feature so if you can start writing the documentation it would be great.

**Comment by simaoribeiro** [comment]
Sure. I will start working on the documentation tomorrow. Please tell me when I can merge your changes on my repo.

**Comment by joao-r-reis** [comment]
There's 2 tests that are failing on appveyor but they didnt fail before, I'm confused. I'll take a look at them tomorrow.

**Comment by joao-r-reis** [comment]
I think I found the cause of the tests flakyness, [here's the PR ](https://github.com/simaoribeiro/csharp-driver/pull/4)to fix it.

**Comment by simaoribeiro** [comment]
Still failing 🤔 Also taking a look

**Comment by joao-r-reis** [comment]
ok I think I fixed it for good now, just running tests on my branch to make sure

**Comment by joao-r-reis** [comment]
[PR ](https://github.com/simaoribeiro/csharp-driver/pull/5)

**Comment by simaoribeiro** [comment]
I think the value of exportedActivities may be empty sometimes as the tests can be fetching the Activities before they closed. Maybe we should implement a retry mechanism in `GetActivities(DateTime from))`

**Comment by simaoribeiro** [comment]
https://github.com/datastax/csharp-driver/pull/608#issuecomment-2348691563
Yep, that's what I was implementing on my side too but yours looks more complete

**Comment by joao-r-reis** [comment]
Yeah, I thought fixing the await on the ResultHandler would guarantee that we have the activities but the activity exporter probably runs on a separate context so there's still a small chance we won't have all the activities when we retrieve them in the test

**Comment by joao-r-reis** [comment]
Ok they[ passed ](https://ci.appveyor.com/project/DataStax/csharp-driver/builds/50596010), you can merge my PR and let's see if this PR's test run is finally green 😅

**Comment by simaoribeiro** [comment]
Merged. Hopefully this will be the last ride 😂

**Comment by joao-r-reis** [comment]
Looks good now 👍 all that's left is documentation and then we can merge this

**Comment by simaoribeiro** [comment]
@joao-r-reis Included documentation in this commit
https://github.com/datastax/csharp-driver/pull/608/commits/d9c697ba4a1a8a90dab708f61fd817aa8eae5073

I also included another commit to null check the `request.Statement` that was missing
https://github.com/datastax/csharp-driver/pull/608/commits/f5585cd6210129f9c5a77659b79ca33f1a4aeb01

**Comment by joao-r-reis** [comment]
Looks good, I added a comment about null pointer exceptions, I think there's still a chance that they happen, it would be nice to have a unit test or something to validate this.

Can you also add a docs page on the new `Request Tracker` interface?

**Comment by simaoribeiro** [comment]
Trying to finish until the end of the day

**Comment by joao-r-reis** [comment]
Oh also apparently `StartActivity` can also return `null` so this can also throw a NPE:

```
 if (activity.IsAllDataRequested)
 {
     if (!string.IsNullOrEmpty(request.Statement.Keyspace))
     {
         activity.AddTag("db.name", request.Statement.Keyspace);
     }

     if (_instrumentationOptions.IncludeDatabaseStatement && request.Statement != null)
     {
         activity.AddTag("db.statement", request.Statement.ToString());
     }
 }
 ```

**Comment by joao-r-reis** [comment]
I'm currently trying this out with a local app and I just realized that the extension method should probably be changed from `AddOpenTelemetryInstrumentation` to `WithOpenTelemetryInstrumentation` because that's what almost all the other builder methods are named like.

**Comment by joao-r-reis** [comment]
I was actually able to reproduce a scenario where the `Activity` is `null` so we should just return right away if that happens and we need to add checks to other methods because the activity might not exist on the context dictionary.

**Comment by joao-r-reis** [comment]
```
Activity.Tags:
    db.system: cassandra
    db.operation: Session Request
    db.statement: Cassandra.BoundStatement
```
The statement CQL string is not being added, just the type of the statement. It's not a blocker for this PR though, it requires some changes to the statement classes so I'll work on this as a follow up PR to this, after this one is merged.

**Comment by joao-r-reis** [comment]
Actually don't worry about the `AddOpenTelemetryInstrumentation` rename either, I'll handle that one too, let's just make sure we're checking if the `Activity` is `null` then I'll merge the PR and work on a follow up PR to take care of these 2 things I brought up.

**Comment by simaoribeiro** [comment]
Regarding the extension method name, `WithOpenTelemetry` is more in line with the rest of Builder methods so makes perfectly sense to change it

**Comment by simaoribeiro** [comment]
The CQL string is included in some use cases given that the test `AddOpenTelemetry_WithIncludeDatabaseStatementOption_DbStatementIsIncludedAsAttribute` is asserting the value of the activity tag against the query that is included in `session.Execute(statement);`

So, maybe we are lacking some other use cases

**Comment by joao-r-reis** [comment]
> The CQL string is included in some use cases given that the test `AddOpenTelemetry_WithIncludeDatabaseStatementOption_DbStatementIsIncludedAsAttribute` is asserting the value of the activity tag against the query that is included in `session.Execute(statement);`
> 
> So, maybe we are lacking some other use cases

Yeah it works with SimpleStatements but not BatchStatements or BoundStatements, I already have a fix on my branch though so I'll merge that once this PR is merged

**Review by joao-r-reis — APPROVED** [review]
Awesome stuff! I'll merge this when CI finishes assuming it's green

**Comment by simaoribeiro** [comment]
Nice. It's a big PR but a good foundation for the next steps. Thanks for all the guidance and help.

There are still some [other attributes](https://opentelemetry.io/docs/specs/semconv/database/cassandra/) that can be included or some behavior that may need to be fixed. I'm available to help if needed.

**Comment by joao-r-reis** [comment]
Yeah, changing the attributes should be simple to do going forward (and don't constitute a breaking change) so even if they aren't 100% perfect on the first release it's not a big deal.

**Comment by joao-r-reis** [comment]
I'll tag you on my follow up PR but it's mostly a justfyi in case you're interested knowing about the changes I make (should be very small though)

**Comment by simaoribeiro** [comment]
Absolutely. Cheers!

**Pull request merged**
